firestore/bulkwriter.go

```diff
@@ -0,0 +1,521 @@
+func (bw *BulkWriter) Set(doc *DocumentRef, datum interface{}, opts ...SetOption) (*WriteResult, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newSetWrites(datum, opts)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot set %v on doc %v", datum, doc.ID)
+	}
+
+	ws, ec := bw.write(w[0])
+	wr, err := bw.processResults(ws, ec)
+	return wr, err
+}
+
+// Update adds a document update write to the queue of writes to send.
+func (bw *BulkWriter) Update(doc *DocumentRef, updates []Update, preconds ...Precondition) (*WriteResult, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newUpdatePathWrites(updates, preconds)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot update doc %v", doc.ID)
+	}
+	wu, ec := bw.write(w[0])
+	wr, err := bw.processResults(wu, ec)
+	return wr, err
+}
+
+// checkConditions determines whether this write attempt is valid. It returns
+// an error if either the BulkWriter has already been closed or if it
+// receives a nil document reference.
+func (bw *BulkWriter) checkWriteConditions(doc *DocumentRef) error {
+	if !bw.isOpen {
+		return errors.New("firestore: BulkWriter has been closed")
+	}
+
+	if doc == nil {
+		return errors.New("firestore: nil document contents")
+	}
+
+	_, loaded := bw.docUpdatePaths.LoadOrStore(doc.shortPath, struct{}{})
+	if loaded {
+		return fmt.Errorf("firestore: bulkwriter: received duplicate write for path: %v", doc.shortPath)
+	}
+
+	return nil
+}
+
+// write packages up write requests into bulkWriterJob objects.
+func (bw *BulkWriter) write(w *pb.Write) (chan *pb.WriteResult, chan error) {
+	r := make(chan *pb.WriteResult, 1)
+	e := make(chan error, 1)
+
+	j := bulkWriterJob{
+		result: r,
+		write:  w,
+		err:    e,
+	}
+
+	// Write successfully created, increment count
+	bw.mu.Lock()
+	wp, ok := bw.writesLog.Load(writesProvided)
+	wpr := wp.(int) + 1
+	if !ok {
+		// TODO(telpirion): Decide what to do
+	}
+	bw.writesLog.Store(writesProvided, wpr)
+	bw.mu.Unlock()
+
+	bw.received <- j
+	return r, e
+}
+
+// processResults checks for errors returned from send() and packages up the
+// results as WriteResult objects
+func (bw *BulkWriter) processResults(w chan *pb.WriteResult, e chan error) (*WriteResult, error) {
+	wpb := <-w
+	err := <-e
+
+	if err != nil {
+		return nil, err
+	}
+
+	wr, err := writeResultFromProto(wpb)
+	if err != nil {
+		return nil, err
+	}
+	return wr, nil
+}
+
+// receiver gets write requests from the caller and sends BatchWriteRequests to the scheduler.
+// The receiver method is the main "event loop" of the BulkWriter. It maintains
+// the communication routes in between the caller and calls to the service.
+func (bw *BulkWriter) receiver(ctx context.Context) {
+	for {
+		var flushOp bool
+		var bs []bulkWriterBatch
+
+		select {
+		case bwj := <-bw.received:
+			bw.sendingQueue = append(bw.sendingQueue, bwj)
+			if bw.isFlushing {
```

Comment:
This looks like a race condition as other goroutines will be changing isFlushing. Can we just keep appending to a new slice of write requests at this point?

Reply:
[Outdated]